client_sdk/go/pkg/fab/ccpackager/packager.go

```diff
@@ -0,0 +1,334 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+/*
+Notice: This file is based on fabric-sdk-go/pkg/fab/ccpackager/lifecycle/packager.go and has been modified.
+*/
+
+// Package ccpackager provides functionality to package a FPC chaincode.
+//
+// Example:
+//  desc := &ccpackager.Descriptor{
+//  	Path:    "/my_fpc_chaincode/build/_lib",
+//  	Type:    ccpackager.ChaincodeType,
+//  	Label:   "my-fpc-chaincode-v1",
+//  	SGXMode: "SIM",
+//  }
+//  ccPkg, err := ccpackager.NewCCPackage(desc)
+//  if err != nil {
+//  	log.Fatal(err)
+//  }
+//
+package ccpackager
+
+import (
+	"archive/tar"
+	"bytes"
+	"compress/gzip"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+	"strings"
+
+	"github.com/hyperledger-labs/fabric-private-chaincode/client_sdk/go/pkg/sgx"
+	"github.com/hyperledger-labs/fabric-private-chaincode/internal/utils"
+	"github.com/hyperledger/fabric/core/chaincode/persistence"
+	"github.com/hyperledger/fabric/core/chaincode/platforms/util"
+	"github.com/pkg/errors"
+)
+
+const (
+	codePackageName          = "code.tar.gz"
+	metadataPackageName      = "metadata.json"
+	connectionsName          = "connection.json"
+	mrenclaveFileName        = "mrenclave"
+	enclaveBinaryName        = "enclave.signed.so"
+	gzipCompressionLevel     = gzip.DefaultCompression
+	ChaincodeType            = "fpc-c"
+	CaaSType                 = "external"
+	defaultConnectionTimeout = "10s"
+)
+
+// NewCCPackage creates a FPC chaincode package.
+func NewCCPackage(desc *Descriptor) ([]byte, error) {
+	err := desc.validate()
+	if err != nil {
+		return nil, err
+	}
+
+	pkgTarGzBytes, err := getTarGzBytes(desc, writePackage)
+	if err != nil {
+		return nil, err
+	}
+
+	return pkgTarGzBytes, nil
+}
+
+// ReadMrenclave returns mrenclave for the given FPC chaincode at ccPath
+func ReadMrenclave(ccPath string) (string, error) {
+	mrenclave, err := ioutil.ReadFile(filepath.Join(ccPath, mrenclaveFileName))
+	if err != nil {
+		return "", err
+	}
+
+	return strings.TrimSuffix(string(mrenclave), "\n"), nil
+}
+
+// Descriptor holds the package data. FPC supports two types of packages, ChaincodeType and CaaSType.
+// For normal chaincode deployments, the package type ChaincodeType is used. It requires to define Type, Label, Path, and SGXMode.
+// Alternatively, for deployments as Chaincode as a Service (CaaS), the package type CaaSType is used.
+// It requires to define Type, Label, Path, and CaaSEndpoint. Optionally, CaaSTimeout and CaaSUseTLS can be set.
+type Descriptor struct {
```

Comment:
Curious why you merged the Descriptor. I guess you document in top-level comment which type needs which descriptor.  However, wouldn't it have been a cleaner way to define a base type with the common fields and two extension types, one for each deployment type?  In that case, you actually wouldn't even have to explicitly add the `Type` field as it would be defined by the type?

Reply:
oh true, I agree that the merged Descriptor is not the cleanest solution. Let me revisit this in the followup PR. My main motivation was just have a single `NewCCPackage` function in our API. But since the CaaSPackage might also work for non FPC chaincode, in particluar, it is also used for packaging ERCC, having a separate function for CaaS might be better.

Reply:
But in case you want CaaS more general, then i think the old approach with different Descriptor types was better as the descriptor is not really general?  Anyway, something for some future discussion ...